### PR TITLE
Add CPU HNSW structural invariant checks with tests

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds a dedicated structural invariant checking framework for the CPU HNSW graph, with tests validating both healthy and corrupted graphs.
- Exposes an invariants API via CpuHnsw, enabling health checks of the graph (layer consistency, degree bounds, reachability, and bidirectional links).
- Adds core utilities (Graph.capacity and Graph.nodes_iter) and per-node helpers (Node.level_count and Node.iter_neighbours) to support invariant evaluation and testing.
- Enhances API surface and documentation to reflect the invariant checker, and provides comprehensive tests covering valid graphs and corruption scenarios.

### Core
- Implemented a modular invariants framework across chutoro-core/src/hnsw/invariants:
  - HnswInvariant: enumerates structural invariants (LayerConsistency, DegreeBounds, Reachability, BidirectionalLinks).
  - HnswInvariantViolation: detailed error types for invariant failures (with payloads like origin, target, layer, and context).
  - HnswInvariantChecker: high-level API to run invariants via check_all, check_many, and individual checks (layer_consistency, degree_bounds, reachability, bidirectional_links).
  - LayerValidator: internal helper to validate node presence and layer availability.
  - Invariant checkers for each invariant and helpers (bidirectional, degree_bounds, layer_consistency, reachability) and tests.
- chutoro-core/src/hnsw/invariants/mod.rs exposed and orchestrated the invariants through a single surface.

- chutoro-core/src/hnsw/graph.rs
  - Added capacity() to return allocated slot count (used for sizing buffers).
  - Added nodes_iter() to iterate over all inserted nodes along with their identifiers.

- chutoro-core/src/hnsw/node.rs
  - Added level_count() to return the number of levels initialised for a node.
  - Added iter_neighbours() to iterate over neighbours as (level, id) tuples.

### API surface
- chutoro-core/src/hnsw/mod.rs
  - Included invariants module so invariants-related types are discoverable.
- chutoro-core/src/lib.rs
  - Re-exported HnswInvariant, HnswInvariantChecker, HnswInvariantViolation for CPU HNSW when the cpu feature is enabled.
- chutoro-core/src/hnsw/mod.rs (CpuHnsw API)
  - Added invariants() -> HnswInvariantChecker<'_> to obtain a handle for invariant checks.
  - Exposed helper methods: layer_consistency, degree_bounds, reachability, bidirectional_links via the checker.

### Documentation
- docs/chutoro-design.md
  - Documented the new structural invariant checkers, their semantics, and how they integrate with property-based tests.
- docs/roadmap.md
  - Updated roadmap to reflect the invariants surface as implemented.

### Tests
- Extended test coverage with a dedicated invariants module tests, including:
  - check_all_succeeds_for_valid_index
  - layer_consistency_reports_invalid_reference (missing node / missing layer)
  - degree_bounds_detects_overflow
  - reachability_fails_for_disconnected_node
  - bidirectional_detects_missing_backlink

## Usage
- Obtain a guard for invariant checks and run a suite or individual checks:
  - Example: index.invariants().check_all() to validate the entire invariants set.
  - Example: index.invariants().check(HnswInvariant::DegreeBounds) to validate a single invariant.
  - You can also run a subset in order using check_many([...] ).

Example snippet:
```rust
let params = HnswParams::new(4, 8).expect("params");
let index = CpuHnsw::build(&DummyVec, params).expect("build");
index.invariants().check_all().expect("graph must be healthy");
```

## Why this matters
- Adds a formalized, testable surface for verifying the structural integrity of the HNSW graph, enabling quick detection of graph corruption during mutations and property-based testing.
- The invariants provide actionable diagnostics with precise context (origin, target, layer, and missing layer details) to streamline debugging and shrink failures during testing.

## Test plan
- [x] Valid graph passes all invariants.
- [x] LayerConsistency detects missing node and missing layer cases.
- [x] DegreeBounds detects overflow at specific layers.
- [x] Reachability detects disconnected nodes starting from the entry point.
- [x] BidirectionalLinks detects missing reverse links.

## How to run tests
- Run cargo test with the cpu feature enabled (as per project conventions):
  - cargo test --features cpu

🌿 Generated by [Terry](https://www.terragonlabs.com)

---
ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/009164b6-29cf-4ec4-93e3-4b39e73fc8cb

## Summary by Sourcery

Add a formal structural invariant checker for the CPU HNSW index with a public API, utility methods, documentation updates, and comprehensive tests to ensure and diagnose graph integrity.

New Features:
- Introduce a structural invariant checking framework for the CPU HNSW graph with definable invariants and detailed violation reporting
- Expose a new invariants API on CpuHnsw via `invariants()` with methods for running all, many, or individual invariant checks
- Add Graph and Node helper methods (`capacity`, `nodes_iter`, `level_count`, `iter_neighbours`) to support invariant evaluation and testing

Enhancements:
- Replace manual candidate iteration with `Vec::contains` in the insertion executor for clarity

Documentation:
- Document the structural invariants and usage of the invariant checker in chutoro-design.md and update the roadmap to include the invariants surface

Tests:
- Add unit tests covering valid HNSW graphs and various corrupted scenarios (missing nodes/layers, degree overflow, unreachable nodes, missing backlinks)

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/hnsw/invariants.rs](https://github.com/leynos/chutoro/pull/36/files/1deb62b74b171ce76a978026483855424bd64b34#diff-ce17ef9ffd278b379c70b7175f741d917619b3385b0bfd5fc50e4494ceabc659)

Comment on lines +333 to +373

~~~rust
fn check_reachability(graph: &Graph) -> Result<(), HnswInvariantViolation> {
    if graph.nodes_iter().next().is_none() {
        return Ok(());
    }
    let entry = graph
        .entry()
        .ok_or(HnswInvariantViolation::MissingEntryPoint)?;
    let validator = LayerValidator::new(graph);
    validator.ensure(entry.node, entry.node, entry.level)?;

    let mut visited = vec![false; validator.capacity()];
    let mut queue = VecDeque::new();
    queue.push_back(entry.node);
    visited[entry.node] = true;

    while let Some(node_id) = queue.pop_front() {
        let node = graph
            .node(node_id)
            .ok_or(HnswInvariantViolation::LayerConsistency {
                origin: node_id,
                target: node_id,
                layer: 0,
                detail: LayerConsistencyDetail::MissingNode,
            })?;
        for (level, target) in node.iter_neighbours() {
            validator.ensure(node_id, target, level)?;
            if visited[target] {
                continue;
            }
            visited[target] = true;
            queue.push_back(target);
        }
    }

    for (node_id, _) in graph.nodes_iter() {
        if !visited[node_id] {
            return Err(HnswInvariantViolation::UnreachableNode { node: node_id });
        }
    }
    Ok(())
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=36&biomarker=Bumpy+Road+Ahead&filename=chutoro-core%2Fsrc%2Fhnsw%2Finvariants.rs&method=check_reachability)
check_reachability has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/hnsw/invariants.rs](https://github.com/leynos/chutoro/pull/36/files/5d0e1e5f864f6780fb09f179cb77dd76f623575a#diff-ce17ef9ffd278b379c70b7175f741d917619b3385b0bfd5fc50e4494ceabc659)

Comment on lines +372 to +388

~~~rust
fn process_neighbours(
    node: &Node,
    validator: LayerValidator<'_>,
    origin: usize,
    visited: &mut [bool],
    queue: &mut VecDeque<usize>,
) -> Result<(), HnswInvariantViolation> {
    for (level, target) in node.iter_neighbours() {
        validator.ensure(origin, target, level)?;
        if visited[target] {
            continue;
        }
        visited[target] = true;
        queue.push_back(target);
    }
    Ok(())
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=36&biomarker=Excess+Number+of+Function+Arguments&filename=chutoro-core%2Fsrc%2Fhnsw%2Finvariants.rs&method=process_neighbours)
process_neighbours has 5 arguments, max arguments = 4

File: chutoro-core/src/hnsw/invariants.rs [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+    ///     .expect("invariant must hold");
+    /// ```
+    pub fn check_many(
+        &self,
+        invariants: impl IntoIterator<Item = HnswInvariant>,
+    ) -> Result<(), HnswInvariantViolation> {
+        for invariant in invariants {
+            self.check(invariant)?;
+        }
+        Ok(())
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/invariants.rs]: Has this now been resolved?

File: chutoro-core/src/hnsw/invariants.rs [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+    fn with_graph<R>(
+        &self,
+        f: impl FnOnce(&Graph, &HnswParams) -> Result<R, HnswInvariantViolation>,
+    ) -> Result<R, HnswInvariantViolation> {
+        let guard = self.index.graph.read().expect("graph lock poisoned");
+        f(&guard, &self.index.params)
+    }
+}
+
+fn check_layer_consistency(graph: &Graph) -> Result<(), HnswInvariantViolation> {
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/invariants.rs]: Has this now been resolved?

File: chutoro-core/src/hnsw/invariants.rs [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+        assert_eq!(
+            err,
+            HnswInvariantViolation::MissingBacklink {
+                origin: 0,
+                target: 1,
+                layer: 0,
+            }
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/invariants.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/hnsw/invariants/reachability.rs](https://github.com/leynos/chutoro/pull/36/files/8c983c9ee2ad950057dc3dcd72d80c5c6bc9eac2#diff-9e8710fa77d16aeb34d17057380c390eeb592cb02052cb3da217419a54c1e375)

Comment on lines +5 to +52

~~~rust
pub(super) fn check_reachability(
    ctx: GraphContext<'_>,
    mode: &mut EvaluationMode<'_>,
) -> Result<(), HnswInvariantViolation> {
    if ctx.graph.nodes_iter().next().is_none() {
        return Ok(());
    }

    let entry = match ctx.graph.entry() {
        Some(entry) => entry,
        None => return mode.record(HnswInvariantViolation::MissingEntryPoint),
    };

    let validator = LayerValidator::new(ctx.graph);
    if let Err(err) = validator.ensure(entry.node, entry.node, entry.level) {
        return mode.record(err);
    }

    let mut context = BfsContext::new(validator.capacity());
    context.visit(entry.node);

    while let Some(node_id) = context.queue.pop_front() {
        let node = ctx
            .graph
            .node(node_id)
            .ok_or(HnswInvariantViolation::LayerConsistency {
                origin: node_id,
                target: node_id,
                layer: 0,
                detail: super::LayerConsistencyDetail::MissingNode,
            })?;

        for (level, target) in node.iter_neighbours() {
            match validator.ensure(node_id, target, level) {
                Ok(_) if context.visited[target] => continue,
                Ok(_) => context.visit(target),
                Err(err) => mode.record(err)?,
            }
        }
    }

    for (node_id, _) in ctx.graph.nodes_iter() {
        if !context.visited[node_id] {
            mode.record(HnswInvariantViolation::UnreachableNode { node: node_id })?;
        }
    }
    Ok(())
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=36&biomarker=Bumpy+Road+Ahead&filename=chutoro-core%2Fsrc%2Fhnsw%2Finvariants%2Freachability.rs&method=check_reachability)
check_reachability has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Bumpy Road Ahead
chutoro-core/src/hnsw/invariants/reachability.rs: bfs_traverse

### What lead to degradation?
bfs_traverse has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

### Why does this problem occur?
A Bumpy Road is a function that contains multiple chunks of nested conditional logic inside the same function. The deeper the nesting and the more bumps, the lower the code health.
A bumpy code road represents a lack of encapsulation which becomes an obstacle to comprehension. In imperative languages there’s also an increased risk for feature entanglement, which leads to complex state management. CodeScene considers the following rules for the code health impact: 1) The deeper the nested conditional logic of each bump, the higher the tax on our working memory. 2) The more bumps inside a function, the more expensive it is to refactor as each bump represents a missing abstraction. 3) The larger each bump – that is, the more lines of code it spans – the harder it is to build up a mental model of the function. The nesting depth for what is considered a bump is  levels of conditionals.

### How to fix it?
Bumpy Road implementations indicate a lack of encapsulation. Check out the detailed description of the [Bumpy Road code health issue](https://codescene.com/blog/bumpy-road-code-complexity-in-context/).
A Bumpy Road often suggests that the function/method does too many things. The first refactoring step is to identify the different possible responsibilities of the function. Consider extracting those responsibilities into smaller, cohesive, and well-named functions. The [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring is the primary response.

**Comment by leynos** [comment]
Please ensure that `make check-fmt`, `make lint` and `make test` all succeed, then commit your changes.@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/hnsw/helpers.rs](https://github.com/leynos/chutoro/pull/36/files/0c546c4258694205a6f44a5a2021a393ad76edfd#diff-7bacd68adf11303a7be20246b17d53d2b21373e061fcf85593a25e2f20e31fa0)

Comment on lines +148 to +162

~~~rust
    fn ensure_query_added_when_room_available() {
        let mut neighbours = vec![neighbour(1, 1.0)];
        let source = TestSource::new(vec![0.0, 1.0]);
        ensure_query_present(
            &cache(),
            EnsureQueryArgs {
                source: &source,
                query: 0,
                ef: NonZeroUsize::new(2).expect("ef must be non-zero"),
                neighbours: &mut neighbours,
            },
        )
        .expect("ensure_query_present must succeed");
        assert_eq!(neighbours, vec![neighbour(0, 0.0), neighbour(1, 1.0)]);
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=36&biomarker=Code+Duplication&filename=chutoro-core%2Fsrc%2Fhnsw%2Fhelpers.rs&method=)
The module contains 3 functions with similar structure: tests.ensure_query_added_when_room_available,tests.ensure_query_noop_when_present,tests.ensure_query_skips_when_capacity_is_one

File: chutoro-core/src/hnsw/invariants/reachability.rs [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        Self {
+            visited: vec![false; capacity],
+            queue: VecDeque::new(),
+        }
+    }
+
+    fn visit(&mut self, node: usize) {
+        self.visited[node] = true;
+        self.queue.push_back(node);
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/invariants/reachability.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: chutoro-core/src/chutoro.rs [hunk 5]
```diff
@@ -255,3 +279,57 @@ impl Chutoro {
...
+            ExecutionStrategy::Auto,
+        );
+        assert!(chutoro.backend_unavailable_error().is_some());
+
+        let cpu_only = Chutoro::new(
+            NonZeroUsize::new(1).expect("literal 1 is non-zero"),
+            ExecutionStrategy::CpuOnly,
+        );
+        assert!(cpu_only.backend_unavailable_error().is_some());
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/chutoro.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: chutoro-core/src/hnsw/helpers.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+///     EnsureQueryArgs {
+///         source: &Dummy,
+///         query: 0,
+///         ef: NonZeroUsize::new(2).unwrap(),
+///         neighbours: &mut neighbours,
+///     },
+/// )?;
+/// assert_eq!(neighbours.len(), 1);
+/// # Ok::<(), chutoro_core::hnsw::error::HnswError>(())
+/// ```
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.
